### ISSUE TITLE
cgroups support in multiprocessing

### ISSUE THREAD

**Issue by 7dc75db9-ae40-4dbb-acc3-2ae6c0db781a** [python/cpython#70879]
BPO | [26692](https://bugs.python.org/issue26692)
--- | :---
Nosy | @pitrou, @giampaolo, @mihaic, @prehensilecode

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2016-04-05.00:46:11.616>
labels = ['3.7', 'type-feature', 'library']
title = 'cgroups support in multiprocessing'
updated_at = <Date 2018-01-16.20:06:33.176>
user = 'https://bugs.python.org/SatrajitGhosh'
```

bugs.python.org fields:
```python
activity = <Date 2018-01-16.20:06:33.176>
actor = 'hairygristle'
assignee = 'none'
closed = False
closed_date = None
closer = None
components = ['Library (Lib)']
creation = <Date 2016-04-05.00:46:11.616>
creator = 'Satrajit Ghosh'
dependencies = []
files = []
hgrepos = []
issue_num = 26692
keywords = []
message_count = 4.0
messages = ['262881', '298893', '298901', '310113']
nosy_count = 8.0
nosy_names = ['pitrou', 'giampaolo.rodola', 'jnoller', 'neologix', 'mihaic', 'sbt', 'Satrajit Ghosh', 'hairygristle']
pr_nums = []
priority = 'normal'
resolution = None
stage = 'needs patch'
status = 'open'
superseder = None
type = 'enhancement'
url = 'https://bugs.python.org/issue26692'
versions = ['Python 3.7']
```

</p></details>

**Comment by 79528080-9d85-4d18-8a2a-8b1f07640dd7** [comment]
I'm not convinced.
The reason is that using the number of CPU cores is just a heuristic
for a \*default value*: the API allows the user to specify the number
of workers to use, so it's not really a limitation.

The problem is that if you try to think about a more "correct" default
value, it gets complicated: here, it's about cgroups, but for example:
- What if they are multiple processes running on the same box?
- What if the process is subject to CPU affinity? Currently, the CPU
affinity mask is ignored.
- What if the code being executed by children is itself multi-threaded
(maybe because it's using a numerical library using BLAS etc)?
- What about hyper-threading? If the code has a lot of cache misses,
it would probably be a good idea to use one worker per logical thread,
but if it's cache-friendly, probably not.
- Etc.

In other words, I think that there's simply not reasonable default
value for the number of workers to use, that any value will make some
class of users/use-case unhappy, and it would add a lot of unnecessary
complexity.

Since the user can always specify the number of workers - if you find
a place where it's not possible, then please report it - I really
think we should let the choice/burden up to the user.

**Comment by pitrou** [comment]
Agreed that it is not possible for multiprocessing to choose an optimal default in all settings.  However, making the default adequate for more use cases sounds like a reasonable goal.

Currently, we are using `os.cpu_count()`. Ideally, we would have a second API `os.usable_cpu_count()` that would return the number of logical CPUs usable by the current process (taking into account affinity settings, cgroups, etc.).

**Comment by zerothi** [comment]
I completely agree on enforcing `cgroups`, it would at least be useful to access num-cgroups processors consistently.

**Comment by vstinner** [comment]
I close this issue as a duplicate of issue #80235.
